designs/karpenter-dra-kwok-driver.md

```diff
@@ -0,0 +1,135 @@
+
+## Implementation
+
+### Case 1: Node Overlay Integration
+Tests **Karpenter's integrated DRA scheduling** where DRA device counts are known during the scheduling simulation via extended resources. NodeOverlay informs Karpenter about expected DRA device capacity during scheduling through extended resources, Provisioner includes these extended resources in NodeClaim templates, and Karpenter provisions nodes knowing they will have specific device counts. The driver then creates ResourceSlices with detailed device information matching the NodeOverlay's extended resource count.
+
+**Example Node Overlay with DRA** (future API extension):
+```yaml
+apiVersion: karpenter.sh/v1alpha1
+kind: NodeOverlay
+metadata:
+  name: gpu-dra-config
+spec:
+  weight: 10  # Higher weight for conflict resolution
+  requirements:
+  - key: node.kubernetes.io/instance-type
+    operator: In
+    values: ["g5.48xlarge"]
+  capacity:
+    fake-gpu.kwok.x-k8s.io/device: "8"  # Custom extended resource for DRA devices
+  # TODO: Extend NodeOverlay API to embed ResourceSlice templates
+  resourceSlices:  # FUTURE: Embedded ResourceSlice objects (not yet implemented)
+  - apiVersion: resource.k8s.io/v1alpha3
+    kind: ResourceSlice
+    spec:
+      # nodeName will be filled in by driver when node is created
+      driver: "fake-gpu.kwok.x-k8s.io"
+      devices:
+      - name: "nvidia-h100-0"
+        driver: "fake-gpu.kwok.x-k8s.io"
+        attributes:
+          memory: "80Gi"
+          compute-capability: "9.0"
+          vendor: "nvidia"
+      - name: "nvidia-h100-1"
+        driver: "fake-gpu.kwok.x-k8s.io"
+        attributes:
+          memory: "80Gi"
+          compute-capability: "9.0"
+          vendor: "nvidia"
+      # ... (6 more devices for total of 8)
+```
+
+**How it works**:
+1. **Test author defines NodeOverlay configuration**: "g5.48xlarge KWOK nodes should have 8x fake H100 GPUs" via ResourceSlices
+2. **Driver watches for KWOK nodes**: When Karpenter creates a KWOK node with `instance-type: g5.48xlarge`
+3. **NodeOverlay match found**: Driver checks for NodeOverlay with embedded ResourceSlice objects, finds matching configuration
+4. **Driver creates ResourceSlice**: Acts as fake DRA driver using embedded ResourceSlice objects from NodeOverlay
+5. **Scheduler sees configured devices**: ResourceSlices with fake devices become available for DRA pod scheduling
+6. **Test validation**: Validates that the driver correctly provides DRA resources and enables successful pod scheduling
+
+### Case 2: ConfigMap Fallback Configuration
+Tests **DRA resource provisioning when no NodeOverlay configuration is found**. When the driver cannot find matching NodeOverlay resources with DRA annotations for a KWOK node, it falls back to ConfigMap-based device configuration. This validates that the driver can handle nodes that weren't pre-configured with DRA resources in NodeOverlay and ensures consistent DRA device availability across different provisioning scenarios.
```

Comment:
> This validates that the driver can handle nodes that weren't pre-configured with DRA resources in NodeOverlay

I think it's important to expand on why this matters. Specifically, I think there's one main use-case: testing that Karpenter successfully discovers ResourceSlices and schedules against them, even if they weren't defined on any NodeOverlays.